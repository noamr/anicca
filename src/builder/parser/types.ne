@{%
const moo = require("moo")
const lexer = moo.compile({
  keywords: ['i8', 'u8', 'i16', 'u16', 'i32', 'u32', 'i64', 'u64', 'f32', 'f64', 'i128', 'u128', 'string', 'bool', 'ByteArray'],
  ws: /[ \t]+/,
  word: /[A-Za-z_$][A-Za-z_$0-9]*/,
  symbols: /[\[\],{}:]/
})

%}
@builtin "whitespace.ne"

@lexer lexer

type ->
    singular {% id %}
    | tuple {% id %}
    | dictionary {% id %}

singular ->
    _ %keywords _ {% ([,{value}]) => value %}

tuple ->
    _ "[" _ args _ "]" _ {% ([,,,args]) => ({tuple: args}) %}

args ->
    type
    | _ args _ "," _ type _ {% ([,multi,,,, single]) => [...multi, single] %}

dictionary ->
    _ "{" _ type _ ":" _ type "}" {% ([,,,a,,,,b]) => ({dictionary: [a, b]}) %}