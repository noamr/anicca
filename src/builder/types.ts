export type Primitive = string | number | boolean | null

export type StatementType = 'Const' | 'View' | 'Main' | 'Let' | 'Controller' | 'Slot' | 'Bus' | 'Table'

export interface WithToken {
    $token?: {
        line: number
        col: number,
    }

}
export interface Statement extends WithToken {
    name?: string
    type: StatementType
}

export type PostProcessor = (bundle: Bundle) => Bundle

export interface ConstStatement extends Statement {
    type: 'Const'
    value: Primitive
}

export interface SlotStatement extends Statement {
    type: 'Slot'
    formula: Formula
}
export interface LetStatement extends Statement {
    type: 'Let'
    valueType: Primitive
}

type SingularType = 'u8' | 'u16' | 'u32' | 'u64' | 'u128' | 'i8' | 'i16' | 'i32' | 'i64' | 'i128' | 'f32' | 'f64' | 'string' | 'bool' | 'ByteArray'
export type NativeType = SingularType | NativeTupleType | NativeDictionaryType
interface NativeTupleType {
    tuple: NativeType[]
}
interface NativeDictionaryType {
    dictionary: [NativeType, NativeType]
}

export interface TableStatement extends Statement {
    type: 'Table'
    valueType: NativeType
}

export interface ViewDeclaration extends WithToken {
    type: 'Bind' | 'DOMEvent'
}

export interface BindTarget extends WithToken  {
    type: 'html' | 'attribute' | 'style'
}

export type Formula = WithToken
export interface ReferenceFormula extends Formula {
    $ref: string
}
export interface PrimitiveFormula extends Formula {
    $primitive: Primitive
}

export interface FunctionFormula<Op extends string = string> extends Formula {
    op: Op
    args?: Formula[]
}

export interface BindDeclaration extends ViewDeclaration {
    type: 'Bind'
    src: Formula
    target?: string
    targetType: 'content' | 'attribute' | 'data' | 'style'
}

export interface DOMEventDeclaration extends ViewDeclaration {
    type: 'DOMEvent'
    eventType: string
    actions: DOMEventAction[]
}

export interface DOMEventAction extends WithToken {
    type: 'PreventDefault' | 'Dispatch' | 'StopPropagation'
}

export interface DispatchAction extends DOMEventAction {
    type: 'Dispatch'
    target: string
    event: string
    payload?: any

}
export interface GotoAction extends TransitionAction {
    type: 'Goto'
    target: string
}

export interface ViewRule extends WithToken  {
    type: 'ViewRule'
    selector: string
    declarations: ViewDeclaration[]
}

export interface AppDeclaration extends WithToken  {
    type: 'Use'
    ref: string
}

export interface MainStatement extends Statement {
    type: 'Main'
    declarations: AppDeclaration[]
}

export interface ControllerStatement extends Statement {
    type: 'Controller'
    name: string
    rootState: State
}

export interface Statechart {
    root: State
}

export interface ViewStatement extends Statement {
    type: 'View'
    name: 'string'
    rules: ViewRule[]
}

export interface Transition extends WithToken {
    type: 'Transition'
    event?: string
    condition?: Formula
    actions?: TransitionAction[]
}

export interface TransitionAction {
    type: 'Assign' | 'Dispatch' | 'Goto'
}
export interface AssignTransitionAction {
    type: 'Assign'
    source: Formula
    method?: 'post' | 'put' | 'delete'
    target: Formula
}

export interface State {
    type: 'State' | 'Parallel' | 'Final' | 'History' | 'Initial'
    name: string
    deep?: boolean
    defaultTargets?: string[]
    defaultActions?: TransitionAction[]
    onEntry?: TransitionAction[]
    onExit?: TransitionAction[]
    children: Array<State|Transition>
}

interface NT<Name> {$T: Name}

type ArgumentTypes<F> = F extends (...args: Array<infer A>) => any ? A : never
type ReturnType<F> = F extends (...args: Array<infer A>) => infer R ? R : never

export interface TypedFormula<T> extends Formula {
    $T: T
}

export interface TypedPrimitive<T> extends TypedFormula<T> {
    $primitive: T
}

export interface TypedRef<T> extends TypedFormula<T> {
    $ref: string
}

type MapKeyType<M> = M extends Map<infer K, any> ? K : never
type MapEntryType<M> = M extends Map<infer K, infer V> ? [K, V] : never
type MapValueType<M> = M extends Map<any, infer V> ? V : never

interface SimpleFunctions {
    gt(a: number, b: number): boolean
    gte(a: number, b: number): boolean
    lt(a: number, b: number): boolean
    lte(a: number, b: number): boolean
    eq(a: number|string|null|boolean, b: number|string|null|boolean): boolean
    neq(a: number|string|null|boolean, b: number|string|null|boolean): boolean
    plus(a: number, b: number): number
    minus(a: number, b: number): number
    mult(a: number, b: number): number
    div(a: number, b: number): number
    pow(a: number, b: number): number
    mod(a: number, b: number): number
    bwand(a: number, b: number): number
    bwor(a: number, b: number): number
    bwxor(a: number, b: number): number
    shl(a: number, b: number): number
    shr(a: number, b: number): number
    ushr(a: number, b: number): number
    bwnot(a: number): number
    negate(a: number): number
    sin(a: number): number
    cos(a: number): number
    tan(a: number): number
    atan(a: number): number
    log(a: number): number
    log2(a: number): number
    log10(a: number): number
    acos(a: number): number
    asin(a: number): number
    sqrt(a: number): number
    floor(a: number): number
    ceil(a: number): number
    round(a: number): number
    trunc(a: number): number
    parseInt(a: string, r: number): number
    parseFloat(a: string, r: number): number
    formatNumber(n: number, r: number): string
    now(): number
    uid(): number
    source(): Map<any, any>
    toLowerCase(s: string): string
    toUpperCase(s: string): string
    startsWith(s: string, a: string): boolean
    endsWith(s: string, a: string): boolean
    stringIncludes(s: string, a: string): boolean
    encode(s: string[]): ArrayBuffer
    noop(): null
    table(n: number): any
}

export type Configuration = Set<State>
export type HistoryConfiguration = Map<State, Configuration>

export interface Modus {
    configuration: Configuration
    history: HistoryConfiguration
}

export interface FlatStatechart {
    junctures: Map<Juncture<string>|null, Array<StepResults<string>>>
    events: string[]
    debugInfo?: {}
}

export interface StepResults<M = Modus> {
    condition: Formula
    execution: TransitionAction[]
    modus: M
}

export interface Juncture<M = Modus> {
    event: string | null
    modus: M
}

export type RootType = 'inbox' | 'outbox' | 'idle' | 'staging'
export interface TransformData {
    tables: {[name: string]: number}
    roots: {[name in RootType]?: Formula}
    refs: {[name: string]: Formula}
    outputNames: {[name: string]: number}
    outputs: {[name: string]: TypedFormula<ArrayBuffer>}
    getEventHeader: (event: string, target: string) => number
    debugInfo: any
    views: {
        bindings: Array<{
            view: string
            selector: string
            target?: string
            type: 'attribute' | 'content' | 'data' | 'style',
        }>
        events: Array<{
            view: string
            selector: string
            eventType: string
            preventDefault: boolean
            stopPropagation: boolean
            headers: number[],
        }>,
    }
}

export function tuple<A, B>(a: A, b: B) {
    return [a, b] as [A, B]
}

type IsTuple<T> = T extends any[] ? number extends T['length'] ? false : true : false
export type toArgType<T> = T | ResolveType<T> | toFormula<T>
export type toFormula<T> =
    T extends TypedFormula<infer R> ? T :
    T extends Array<TypedFormula<infer R>> ? TypedFormula<R[]> :
    T extends {$T: infer R} ? TypedFormula<R> :
    TypedFormula<T>
type ValueTypeOf<T, K = any> =
    T extends Map<any, infer V> ? V :
    T extends {[key: number]: infer V} ? V :
    T extends {[key: string]: infer V} ? V :
    IsTuple<T> extends true ? never :
    T extends Array<infer V> ? V :
    never

type KeyTypeOf<T> =
    T extends Map<infer K, any> ? K :
    T extends {[key: number]: infer V} ? number :
    T extends {[key: string]: infer V} ? string :
    keyof T

type KeyType<T> = KeyTypeOf<ResolveType<T>>
type ValueType<T, K = any> = ResolveType<ValueTypeOf<ResolveType<T>, K>>
type ResolveType<P> = P extends {$T: infer T} ? T : P
type IsMapType<T> = ResolveType<T> extends Map<any, any> ? true : never
type Pair<A, B> = [A, B]

export type AssignmentDirective<K = any, V = any> = [number, ResolveType<K>, ResolveType<V>]

export type FormulaBuilder = {
    [k in keyof SimpleFunctions]: (...args: Array<toArgType<ArgumentTypes<SimpleFunctions[k]>>>) =>
        toFormula<ReturnType<SimpleFunctions[k]>>
} & {
    get<M, K>(s: M, k: K): toFormula<ValueType<M, K>>
    first<P>(s: P): ResolveType<P> extends ResolveType<[infer A, any]> ? toFormula<A> : never
    last<P>(s: P): ResolveType<P> extends ResolveType<[any, infer B]> ? toFormula<B> : never
    flatMap<M, P>(input: M, predicate: P):
        IsMapType<M> extends true ? toFormula<P> extends toFormula<Array<[infer K2, infer V2]>> ? toFormula<Map<K2, V2>>
        : never : never
    map<M, P>(input: M, predicate: P):
        IsMapType<M> extends true ? toFormula<P> extends toFormula<infer V2> ? toFormula<Map<KeyType<M>, V2>>
        : never : never
    flatReduce<M, P, V>(map: M, predicate: P, initialValue: V):
        IsMapType<M> extends true ?
            ResolveType<P> extends [boolean, ResolveType<V>] ? toFormula<V>
            : never : never
    head<M>(a: M): toFormula<KeyType<M>>
    tail<M>(a: M): toFormula<KeyType<M>>
    findFirst<T, P>(t: T, p: P): IsMapType<T> extends true ? toFormula<KeyType<T>> : never
    concat<A, B>(a: A, b: B): toFormula<Array<ValueType<A> | ValueType<B>>>
    object<P>(...entries: P[]): P extends toArgType<Pair<infer K, infer V>> ? toFormula<Map<K, V>> : never
    array<V>(...entries: V[]): toFormula<V[]>
    pair<A, B>(a: A, b: B): toFormula<[ResolveType<A>, ResolveType<B>]>
    not<T>(o: any): toFormula<boolean>
    and<A>(...args: A[]): toFormula<A>
    or<A>(...args: A[]): toFormula<A>
    key<T = string|number>(): toFormula<T>
    value<T = any>(): toFormula<T>
    aggregate<T = any>(): toFormula<T>
    size<T>(m: T): toFormula<number>
    isnil<A>(a: A): toFormula<A extends null ? true : boolean>
    cond<Condition, Consequent, Alternate>(c: Condition, t: Consequent, a: Alternate):
        toFormula<Consequent | Alternate>
    put<T, K, V>(table: T, key: K, value: V): toFormula<AssignmentDirective<K, V>>
    delete<T, K>(table: T, key: K): toFormula<AssignmentDirective<ResolveType<K>, any>>
    replace(): toFormula<AssignmentDirective>
    merge(): toFormula<AssignmentDirective>
    filter<T, P>(t: T, p: P): IsMapType<T> extends true ? toFormula<T> : never
    some<T, P>(t: T, p: P): IsMapType<T> extends true ? toFormula<boolean> : never
    every<T, P>(t: T, p: P): IsMapType<T> extends true ? toFormula<boolean> : never
    diff<T>(a: T, b: T): IsMapType<T> extends true ? toFormula<T> : never,
}

export type Bundle = Statement[]

export interface RawFormula {
    op?: string
    args?: number[]
    value?: any
}

export interface StoreSpec {
    roots: {[key in RootType]: number}
    outputNames: {[name: string]: number}
    slots: RawFormula[]
    tableTypes: {[x: number]: NativeType}
    debugInfo?: any[]
}
