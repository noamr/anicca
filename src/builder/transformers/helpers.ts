import { Bundle, Formula, FormulaBuilder, PrimitiveFormula, Statement, WithToken, Token } from '../types'
export const P = ($primitive: any) => ({$primitive}) as PrimitiveFormula

function fixArg(a: any): any {
    if (typeof a === 'undefined') {
        throw new Error('Unexpected undefined')
    }

    if (typeof a !== 'object' || a === null)
        return {$primitive: a}

    if (a.$ref || Reflect.has(a, '$primitive') || Reflect.has(a, '$type'))
        return a

    if (Array.isArray(a)) {
        return {op: 'array', args: a.map(fixArg)}
    }

    if (a.op)
        return {...a, args: (a.args || []).map(fixArg)}

    return {op: 'object', args: Object.entries(a).map(([key, value]) => F.pair(key, value)), $token: a.$token}
}


export function assert<T>(value: T | null | undefined, message?: string): T {
    if (value)
        return value

    if (message)
        throw new Error(message)

    const error = new Error('assertion')
    if (error.stack) {
        const lines = error.stack.split('n')
        if (lines && lines.length > 2) {
            const relevantLine = lines[2]
            const match = relevantLine.match(/(?<=at )\w+ \([^$]+\)/)
            if (match)
                error.message = `Assertion failed at ${match[0]}`
        }
    }
    throw error
}

export function withInfo<T extends Formula = Formula>(f: T, info: string | null): T {
    if (!info)
        return f
    return {...f, $token: {...f.$token, info}}
}

function createToken(error: Error): Partial<Token> | null{
    return (((error.stack || '').split('\n')[2] || '')
    .match(/\((?<file>[^:]+):(?<line>\d+):(?<col>\d+)\)/i) || {groups: null}).groups || null
}

export const F = new Proxy({}, {
    get: (t, op: string) => (...args: any[]) => ({
        op, args: args.map(fixArg), $token: createToken(new Error())
    }),
}) as FormulaBuilder

export const S = new Proxy({}, {
    get: (t, type) => (name: string, stuff?: any) => ({
        type, name, ...stuff,
    }),
}) as {[x: string]: (name: any, stuff?: any) => Statement}

export const R = ($ref: string) => ({$ref})

export const removeUndefined = (a: any) => JSON.parse(JSON.stringify(a))
