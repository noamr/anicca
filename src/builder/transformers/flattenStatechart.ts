import * as _ from 'lodash-es'
import { Bundle, Configuration, ControllerStatement, DispatchAction, FlatStatechart, Formula,
    GotoAction, HistoryConfiguration, Juncture, Modus, State, Statechart, StepResults, Transition, TransitionAction,
 } from '../types'
import {F, R, removeUndefined, S} from './helpers'

const head = <T>(a: T[]) => a.length ? a[0] : null
const flatten = <T>(a: Iterable<Iterable<T>>): Iterable<T> => Array.from(a).reduce((a, b) => [...a, ...b], [])
type Predicate<Ret, Arg> = (a: Arg) => Ret
const some = <T>(a: Iterable<T>, p: Predicate<boolean, T>) => Array.from(a).some(p)
const intersectSets = <T>(a: Set<T>, b: Set<T>) => some(a, x => some(b, y => y === x))
const sort = <T>(a: Iterable<T>, p: (a: T, b: T) => number) => a ? Array.from(a).sort(p) : a

export function flattenState(rootState: State): FlatStatechart {
    const byName: {[name: string]: State} = {}
    const parentMap = new WeakMap<State, State|null>()
    const transitionSourceMap = new WeakMap<Transition, State>()
    const docIndex = new WeakMap<State, number>()
    const allEvents = new Set<string>()

    const getDefault = (state: State): State[] => (state.defaultTargets || []).map(id => byName[id])
    const isParallelState = (s: State) => s.type === 'Parallel'
    const documentIndex = (s: State): number => docIndex.get(s) || 0

    const entryOrder = (a: State, b: State) => documentIndex(a) - documentIndex(b)
    const exitOrder = (a: State, b: State) => documentIndex(b) - documentIndex(a)

    const isAtomicState = (s: State) => s.type === 'State' && !childStates(s).length
    const conditionFor = (t: Transition|null): Formula => (t && t.condition) || trueCondition
    const trueCondition = {$primitive: true} as Formula
    const falseCondition = {$primitive: true} as Formula

    const hashState = (s: State) => `${s.name || `#${documentIndex(s)}`}`
    const hashConfig = (c: Configuration) => sort(c, entryOrder).map(hashState).join(',')
    const hashHistory = (c: HistoryConfiguration) => [...c.keys()].sort(entryOrder).map((hs: State) => `${hashState(hs)}:${hashConfig(c.get(hs) || new Set<State>())}`).join('|')
    const hashModus = (m: Modus): string => `${hashConfig(m.configuration)};${hashHistory(m.history)}`
    const hashJuncture = (j: Juncture|null): string => j ? `${j.event || ''}:${hashModus(j.modus)}` : ''

    const junctures: {[hash: string]: {
        juncture: Juncture|null
        results: StepResults[],
    }} = {}

    const visitedModi = new Set<string>()
    const childStates = (s: State): State[] =>
        ((s.children || []).filter(({type}) => ['State', 'Parallel', 'Final'].includes(type)) as State[])

    const childTransitions = (s: State): Transition[] =>
        ((s.children || []).filter(({type}) => type === 'Transition') as Transition[])

    const getParent = (s: State) => parentMap.get(s) || rootState
    const getTransitionSource = (t: Transition) => transitionSourceMap.get(t) as State

    const nextDocIndex = (next => () => next++)(0)

    resolveState(rootState, null)

    function resolveJuncture(juncture: Juncture|null): void {
        const junctureHash = hashJuncture(juncture)
        if (junctures[junctureHash])
            return

        const results = step(juncture)
        junctures[junctureHash] = {juncture, results}
        for (const r of results) {
            const modusHash = hashModus(r.modus)
            if (visitedModi.has(modusHash))
                continue

            visitedModi.add(modusHash)
            resolveModus(r.modus)
        }
        return
    }

    const resolveModus = (modus: Modus) =>
        [null, ...allEvents].forEach(event => resolveJuncture({event, modus}))

    resolveJuncture(null)
    return {
        events: [...allEvents],
        junctures: new Map(Object.values(junctures).map(({juncture, results}) =>
        ([juncture ? {event: juncture.event, modus: hashModus(juncture.modus)} : null,
            results.map(({condition, execution, modus}) => ({
                condition, execution, modus: hashModus(modus),
            }))] as [Juncture<string>|null, Array<StepResults<string>>]))),
    }

    function resolveTransition(t: Transition, s: State) {
        transitionSourceMap.set(t, s)
        if (t.event)
            allEvents.add(t.event)
    }

    function resolveState(s: State, parent: State|null) {
        byName[s.name] = s
        parentMap.set(s, parent)
        docIndex.set(s, nextDocIndex())

        childStates(s).forEach(child => resolveState(child, s))
        childTransitions(s).forEach(child => resolveTransition(child, s))
    }

    function getAncestors(s1: State): State[] {
        if (!s1 || s1 === rootState)
            return []
        const parent = getParent(s1)
        return [parent, ...getAncestors(parent)].filter(a => a) as State[]
    }

    function getTransitionTargets(t: Transition): State[] {
        const e = (t.actions || []).filter((a: TransitionAction) => a.type === 'Goto')
        return e.map(g => byName[(g as GotoAction).target])
    }

    interface ConditionalTransitionSet {
        condition: Formula
        transitions: Transition[]
    }

    function step(juncture: Juncture|null): StepResults[] {
        if (!juncture) {
            const {modus, execution} = microstep(
                [{type: 'Transition', actions: [{type: 'Goto', target: rootState.name} as GotoAction]}],
                new Set<State>(), new Map())
            return [{
                modus,
                execution,
                condition: trueCondition,
            }]
        }

        const {event, modus: {configuration, history}} = juncture
        const transitionSets = selectTransitions()
        return transitionSets.map(
            t => ({condition: t.condition, ...microstep(t.transitions, configuration, history)}) as StepResults)

        function selectTransitions(): ConditionalTransitionSet[] {
            const atomicStates = [...configuration].filter(isAtomicState).sort(entryOrder)
            const atomicTransitionSets = atomicStates.map(state => {
                const transitions = new Set<Transition>()
                const states = [state, ...getProperAncestors(state, null)]
                for (const s of states)
                    for (const t of childTransitions(s).filter((t => event ? (event === t.event) : !t.event)))
                        transitions.add(t)
                const tlist = [...transitions]
                return tlist.map((t, i) => ({
                    ...t,
                    condition:
                        i === 0 ? conditionFor(t) :
                        i === 1 ? (F.and(F.not(conditionFor(tlist[0])), conditionFor(t))) :
                        F.and(F.not(conditionFor(tlist[0])), conditionFor(t)),
                }) as Transition)
            })

            const computePivot = (current?: Transition[], ...next: Transition[][]): Transition[][] => {
                if (!current)
                    return []
                if (!next.length)
                    return [current]

                return computePivot(...next).map(
                    (p: Transition[]) => [...flatten(current.map(t => ([t, ...p])))] as Transition[])
            }

            const pivot: Transition[][] = computePivot(...atomicTransitionSets)
            return pivot.map((transitions) => ({
                condition:
                    transitions.length === 1 ?
                    conditionFor(head(transitions)) :
                    F.and(...transitions.map(conditionFor)),
                transitions: filterConflicts(transitions),
            }))

            function filterConflicts(selected: Transition[]): Transition[] {
                const filteredTransitions = new Set<Transition>()
                selected.forEach((t1, i) => {
                    let t1Preempted = false
                    const transitionsToRemove = new Set<Transition>()
                    for (const t2 of selected.slice(i + 1)) {
                        if (intersectSets(computeExitSet([t1]), computeExitSet([t2]))) {
                            if (isDescendant(getTransitionSource(t1), getTransitionSource(t2)))
                                transitionsToRemove.add(t2)
                            else
                                t1Preempted = true
                        }
                    }
                    if (!t1Preempted) {
                        for (const t3 of transitionsToRemove)
                            filteredTransitions.delete(t3)
                        filteredTransitions.add(t1)
                    }
                })
                return [...filteredTransitions]
            }
        }

        function getEffectiveTargetStates(t: Transition): State[] {
            let targets: State[] = []
            for (const s of getTransitionTargets(t))
                if (s.type === 'History')
                    if (history.has(s))
                        targets = [...targets, ...(history.get(s) || [])]
                    else
                        targets = [...targets, ...getDefault(s)]
            return [...new Set(targets)]
        }

        function getTransitionDomain(t: Transition) {
            const targets = getEffectiveTargetStates(t)
            const source = getTransitionSource(t)
            if (!targets.length)
                return rootState
            if (isCompound(source) && targets.every(t => isDescendant(t, source)))
                return source
            return findLCCA([source, ...targets]) || rootState
        }

        function computeExitSet(transitions: Transition[]): Set<State> {
            if (!juncture)
                return new Set<State>()

            const statesToExit = new Set<State>()
            transitions
                .filter(t => getTransitionTargets(t).length)
                .forEach(t => {
                    const domain = getTransitionDomain(t)
                    for (const s of configuration)
                        if (!domain || isDescendant(s, domain))
                            statesToExit.add(s)
                })

            return statesToExit
        }

        function microstep( transitions: Transition[],
                            prevConfiguration: Configuration,
                            prevHistory: HistoryConfiguration): {
            modus: Modus,
            execution: TransitionAction[],
        } {
            const configuration = new Set([...prevConfiguration])
            const history = new Map([...prevHistory])
            const onExit = computeStateExit(transitions)
            const onEntry = computeStateEntry(transitions)
            const transitionExec = transitions.map(t => (t.actions || []).filter(a => a.type !== 'Goto'))
                .reduce((a, o) => a.concat(o), [])

            return {
                modus: {configuration, history},
                execution: [...onExit, ...transitionExec, ...onEntry],
            }

            function computeStateExit(transitions: Transition[]): TransitionAction[] {
                let execution: TransitionAction[] = []

                const statesToExit = [...computeExitSet(transitions)].sort(exitOrder)
                for (const s of statesToExit) {
                    for (const h of s.children.filter(c => c.type === 'History') as State[]) {
                        history.set(h, new Set([...configuration].filter(
                            s0 => h.deep
                                ? (isAtomicState(s0) && isDescendant(s0, s))
                                : getParent(s0) === s)))
                    }
                    execution = [...execution, ...(s.onExit || [])]
                    configuration.delete(s)
                }

                return execution
            }

            function computeStateEntry(transitions: Transition[]): TransitionAction[] {
                const statesToEnter = new Set<State>()
                const defaultHistoryContent = new Map<State, TransitionAction[]>()
                const statesForDefaultEntry = new Set<State>()
                let execution: TransitionAction[] = []

                for (const t of transitions) {
                    const targets = getTransitionTargets(t)
                    for (const s of targets)
                        addDescendantStatesToEnter(s)
                    const ancestor = getTransitionDomain(t)
                    for (const s of getEffectiveTargetStates(t))
                        addAncestorStatesToEnter(s, ancestor)
                }

                for (const state of [...statesToEnter].sort(entryOrder)) {
                    configuration.add(state)
                    const onEntry = (state.onEntry || []) as TransitionAction[]
                    const defaultActions = (state.defaultActions || []) as TransitionAction[]
                    execution = [...execution, ...onEntry, ...(statesForDefaultEntry.has(state) ? defaultActions : [])]
                }

                return execution

                function addDescendantStatesToEnter(state: State) {
                    if (state.type === 'History') {
                        const hs = history.get(state)
                        if (hs) {
                            for (const s of hs)
                                addDescendantStatesToEnter(s)
                            for (const s of hs)
                                addAncestorStatesToEnter(s, getParent(s))
                        } else {
                            defaultHistoryContent.set(getParent(state), state.defaultActions || [])
                            for (const s of getDefault(state)) {
                                addDescendantStatesToEnter(s)
                                addAncestorStatesToEnter(s, getParent(state))
                            }
                        }
                        return
                    }
                    statesToEnter.add(state)
                    if (isCompound(state)) {
                        statesForDefaultEntry.add(state)
                        for (const s of getDefault(state)) {
                            addDescendantStatesToEnter(s)
                            addAncestorStatesToEnter(s, state)
                        }
                    } else if (isParallelState(state)) {
                        for (const child of childStates(state))
                            if (!some(statesToEnter, s => isDescendant(s, child)))
                                addDescendantStatesToEnter(child)
                    }
                }

                function addAncestorStatesToEnter(state: State, ancestor: State) {
                    for (const anc of getProperAncestors(state, ancestor)) {
                        statesToEnter.add(anc)
                        if (isParallelState(anc))
                            for (const child of childStates(anc))
                                if (!some(statesToEnter, s => isDescendant(s, child)))
                                    addDescendantStatesToEnter(child)
                    }
                }
            }
        }

        function isCompound(state: State): boolean {
            return state.type === 'State' && !!childStates(state).length
        }

        function isCompoundOrRoot(state: State): boolean {
            return state === rootState || isCompound(state)
        }

        function findLCCA(states: State[]): State {
            if (!states.length)
                return rootState

            for (const anc of getProperAncestors(states[0], null).filter(isCompoundOrRoot)) {
                if (states.slice(1).every(s => isDescendant(s, anc)))
                    return anc
            }

            return rootState
        }

        function isDescendant(s1: State, s2: State) {
            return getAncestors(s1).includes(s2)
        }

        function getProperAncestors(s1: State, s2: State | null) {
            if (s1 === s2)
                return []
            const a = getAncestors(s1)
            if (!s2)
                return a

            const b = getAncestors(s2)
            if (b.indexOf(s1))
                return []

            const s2Index = a.indexOf(s2)
            if (s2Index < 0)
                return a

            return a.slice(0, s2Index)
        }
    }
}
