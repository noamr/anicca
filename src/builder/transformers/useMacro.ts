import * as _ from 'lodash'
import YAML from 'yaml'
import { parseFormula } from '../parser/index'
import { Formula, FunctionFormula, ReferenceFormula } from '../types'
import {F, R, removeUndefined, S} from './helpers'

interface RawMacro {
    inputs: string[]
    output: string
    formulas: {[name: string]: string}
}
export default function useMacro(macroYaml: string, externalFormulas: {[key: string]: any}): Formula {
    const raw = YAML.parse(macroYaml) as RawMacro
    const internalFormulas = _.mapValues(raw.formulas, parseFormula)
    const resolvedFormulas = {...externalFormulas}
    const resolveFormula = (f: Formula): Formula => {
        const {$ref} = f as ReferenceFormula
        if ($ref)
            return resolvedFormulas[$ref] || (
                internalFormulas[$ref] ?
                (resolvedFormulas[$ref] = resolveFormula(internalFormulas[$ref])) : (() => {
                    throw new Error(`Unknown formula: ${$ref}`)
                })())

        const {args} = f as FunctionFormula
        if (args && args.length)
            return {...f, args: args.map((a, i) => a ? resolveFormula(a) : (() => {
                throw new Error(`Null arg. Formula: ${JSON.stringify(f)}, arg: ${i}`)
            })())} as FunctionFormula
        return f
    }

    return resolveFormula({$ref: raw.output} as ReferenceFormula)
}
