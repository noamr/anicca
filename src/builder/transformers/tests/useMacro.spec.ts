import useMacro from '../useMacro'
import {parse} from '../../index'
import { ControllerStatement, ReferenceFormula } from '../../types'
import { FlatStatechart } from '../flattenStatechart'
import {F} from '../helpers'

describe('macros', () => {
    it('basic', () => {
        expect(useMacro(`
        inputs:
            - a
            - b
        output:
            result

        formulas:
            consty: 5.5
            internal: |
                pow(external, consty)
            result: |
                internal * a + b
        `, {
            a: {$primitive: 123},
            b: {$ref: 'bla'},
            external: F.ceil(3)
        })).toMatchSnapshot()
    })
})