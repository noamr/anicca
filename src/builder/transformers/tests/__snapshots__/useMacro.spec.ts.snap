// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros basic 1`] = `
Object {
  "args": Array [
    Object {
      "args": Array [
        Object {
          "args": Array [
            Object {
              "$token": Object {
                "col": "25",
                "file": "/Users/noam/anicca/src/builder/transformers/tests/useMacro.spec.ts",
                "line": "24",
              },
              "args": Array [
                Object {
                  "$primitive": 3,
                },
              ],
              "op": "ceil",
            },
            Object {
              "args": Array [],
              "op": "object",
            },
          ],
          "op": "pow",
        },
        Object {
          "$primitive": 123,
        },
      ],
      "op": "mult",
    },
    Object {
      "$ref": "bla",
    },
  ],
  "op": "add",
}
`;
